debugger/loadprinter.ml

```diff
@@ -90,51 +89,43 @@ let eval_value_path env path =
 
 (* Install, remove a printer (as in toplevel/topdirs) *)
 
-(* since 4.00, "topdirs.cmi" is not in the same directory as the standard
-  library, so we load it beforehand as it cannot be found in the search path. *)
-let init () =
-  let topdirs =
-    Filename.concat !Parameters.topdirs_path "topdirs.cmi" in
-  ignore (Env.read_signature "Topdirs" topdirs)
-
-let match_printer_type desc typename =
-  let printer_type =
-    match
-      Env.find_type_by_name
-        (Ldot(Lident "Topdirs", typename)) Env.empty
-    with
-    | path, _ -> path
-    | exception Not_found ->
-        raise (Error(Unbound_identifier(Ldot(Lident "Topdirs", typename))))
-  in
+let match_printer_type env desc typepath =
   Ctype.begin_def();
   let ty_arg = Ctype.newvar() in
-  Ctype.unify Env.empty
-    (Ctype.newconstr printer_type [ty_arg])
+  Ctype.unify env
+    (Ctype.newconstr typepath [ty_arg])
     (Ctype.instance desc.val_type);
   Ctype.end_def();
   Ctype.generalize ty_arg;
   ty_arg
 
-let find_printer_type lid =
-  match Env.find_value_by_name lid Env.empty with
+let find_printer_type env lid =
+  match Env.find_value_by_name lid env with
   | (path, desc) -> begin
-      match match_printer_type desc "printer_type_new" with
+      let tmp_sign = Genprintval.printertypes_sig env in
+      let tmp_env = Env.add_signature tmp_sign env in
+      let printer_type_new, printer_type_old =
+        Genprintval.printer_types tmp_sign
+      in
+      match match_printer_type tmp_env desc printer_type_new with
       | ty_arg -> (ty_arg, path, false)
       | exception Ctype.Unify _ -> begin
-          match match_printer_type desc "printer_type_old" with
+          match match_printer_type env desc printer_type_old with
           | ty_arg -> (ty_arg, path, true)
           | exception Ctype.Unify _ -> raise(Error(Wrong_type lid))
         end
     end
   | exception Not_found ->
       raise(Error(Unbound_identifier lid))
 
+let local_env = ref Env.empty
+
 let install_printer ppf lid =
-  let (ty_arg, path, is_old_style) = find_printer_type lid in
+  if !local_env = Env.empty then local_env := Compmisc.initial_env ();
```

Comment:
could `local_env` have type `Env.t option ref` to avoid relying on structural equality here?

Reply:
(Or: `let local_env = lazy (Compmisc.initial_env ())`, then Lazy.force in this function.)